core/store/src/trie/mod.rs

```diff
@@ -550,27 +553,48 @@ impl WrappedTrieChanges {
         self.trie_changes.deletions_into(self.trie.clone(), store_update)
     }
 
-    pub fn key_value_changes_into(
-        &self,
+    /// Save state changes into Store.
+    ///
+    /// NOTE: the changes are drained from `self`.
+    pub fn state_changes_into(
+        &mut self,
         store_update: &mut StoreUpdate,
     ) -> Result<(), Box<dyn std::error::Error>> {
         store_update.trie = Some(self.trie.clone());
-        for (key, changes) in &self.kv_changes {
+        for (key, raw_changes) in
+            std::mem::replace(&mut self.state_changes, Default::default()).into_iter()
+        {
             assert!(
-                !changes.iter().any(|(change_cause, _)| {
-                    if let StateChangeCause::NotWritableToDisk = change_cause {
+                !raw_changes.iter().any(|RawStateChange { cause, .. }| {
+                    if let StateChangeCause::NotWritableToDisk = cause {
                         true
                     } else {
                         false
                     }
                 }),
                 "NotWritableToDisk changes must never be finalized."
             );
+            let kind = if let Ok(account_id) = KeyForData::parse_account_id(&key) {
+                StateChangeKind::DataTouched { account_id }
+            } else if let Ok(account_id) = KeyForAccount::parse_account_id(&key) {
+                StateChangeKind::AccountTouched { account_id }
```

Comment:
@evgenykuzyakov I have pushed the StateChangeKind deduction to the stage where we save the changes, so there is no need to parse the keys on RPC query anymore (just use the StateChangeKind, which is now a part of the saved state-change value). It seems that we will be able to resolve `account_id` by hash in this context, right?